pkg/user/external_server.go

```diff
@@ -0,0 +1,263 @@
+package user
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+func readRequestBody(req *http.Request, body interface{}) error {
+	if err := json.NewDecoder(req.Body).Decode(body); err != nil {
+		return err
+	}
+	return nil
+}
+
+func sendResponse(rw http.ResponseWriter, status int, body interface{}) {
+	rw.Header().Set("Content-Type", "application/json")
+	rw.WriteHeader(status)
+
+	if body != nil {
+		// Set up the pipe to write data directly into the Reader.
+		pr, pw := io.Pipe()
+
+		// Write JSON-encoded data to the Writer end of the pipe.
+		// Write in a separate concurrent goroutine, and remember
+		// to Close the PipeWriter, to signal to the paired PipeReader
+		// that we’re done writing.
+		go func() {
+			pw.CloseWithError(json.NewEncoder(pw).Encode(body))
+		}()
+
+		_, err := io.Copy(rw, pr)
+		if err != nil {
+			log.WithError(err).Warn("[ExternalServer] error writing response")
+		}
+	}
+}
+
+func StartExternalServer(listenAddr string, stop <-chan struct{}, um UserManager) error {
```

Comment:
Is this only used in the tests at the moment??

Reply:
I understand gateway would have to import and start this

Reply:
another really bad reason why not to do such things is that it's lacking panic recovery middleware so a panic here would take down the gateway

Reply:
Only in the tests right now, but when the gateway provides a "real" external authz server, it'll use this (it vendors the package from dotscience already)

Reply:
> another really bad reason why not to do such things is that it's lacking panic recovery middleware so a panic here would take down the gateway

I'll pop in a panic catcher!

Reply:
...panic catcher popped in.

Reply:
Consider moving this over to the gateway, for a cleaner separation between dotmesh and the gateway? Would that be too much of a hassle?

Reply:
Well, as the tests in dotmesh use this code, this would require it to be vendored back into gateway, which would probably create circular loops and make dependency management hard - while this package is already vendored into gateway so this way no new imports are needed. Also, keeping the client and server together in one package means that any changes to the protocol can be done in one place, which simplifies keeping client and server in sync!

Reply:
aren't tests using the gomock version of this interface?

Reply:
Nope, it's the real thing! The tests bring up an instance of the external server, and then bring up a Dotmesh cluster and tell it to use the external server, and then poke at the cluster to make it do various things in order to trigger the external server instance created by the tests. Thereby testing the configuration switch in dotmesh to use an external server, the external *client* in Dotmesh, and the external server brought up by the tests.